clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         }
 #ifdef UNIX
-        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT)
-            shard->prev_xfer_int_pc_.push(memref.marker.marker_value);
         report_if_false(shard, memref.marker.marker_value != 0,
                         "Kernel event marker value missing");
-        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT &&
-            // XXX: Handle the back-to-back signals case where the second
-            // signal arrives just after the return from the first without
-            // any intervening instrs. The return point of the second one
-            // would be the pc in the kernel xfer marker of the first.
-            shard->prev_xfer_marker_.marker.marker_type !=
-                TRACE_MARKER_TYPE_KERNEL_XFER) {
-            if (shard->saw_kernel_xfer_after_prev_instr_) {
-                // We have nested signals without an intervening app instr.
-                // Push an empty instr to mean that this shouldn't be used.
-                shard->pre_signal_instr_.push({});
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_XFER) {
+            // We assume paired signal entry-exit (so no longjmp and no rseq
+            // inside signal handlers).
+            if (shard->prev_xfer_int_pc_.empty() ||
+                shard->pre_signal_instr_.size() == 1 ||
+                shard->prev_xfer_aborted_rseq_.empty()) {
+                // If one is true, all should be so.
+                assert(shard->prev_xfer_int_pc_.empty() &&
+                       shard->pre_signal_instr_.size() == 1 &&
+                       shard->prev_xfer_aborted_rseq_.empty());
+                // This can happen if tracing started in the middle of a signal.
+                // Try to continue by skipping the checks.
+                shard->last_xfer_int_pc_ = 0;
+                shard->last_pre_signal_instr_ = {};
+                shard->last_xfer_aborted_rseq_ = false;
+                // Now that we know there's an outer context that we didn't see before,
+                // forget the last instr we saw in what was incorrectly assumed previously
+                // to be the outer-most context.
+                replace_top<memref_t>(shard->pre_signal_instr_, {});
+            } else {
+                shard->last_xfer_int_pc_ = shard->prev_xfer_int_pc_.top();
+                shard->prev_xfer_int_pc_.pop();
+                shard->pre_signal_instr_.pop();
+                // The last instr in the context immediately outer to this signal may
+                // be {} in some cases:
+                // - for nested signals without any intervening instr
+                // - if there's a signal at the very beginning of the trace
+                // In both these cases the empty instr implies that it should not
+                // be used for the pre-signal instr check.
+                shard->last_pre_signal_instr_ = shard->pre_signal_instr_.top();
```

Comment:
Why do some of these 3 stacks do "last=top; pop" while others do "pop; last=top"?  Shouldn't they all be the same wrt whether the top or second slot is what we want?

Reply:
last_pre_signal_instr_ needs to get the second-to-top (the instr in the scope that was interrupted by that signal).

Reply:
I don't follow this: what's at the top then?

I'm thinking we should have a single stack with a compound data structure, rather than a bunch of separate stacks which we have to worry about being out of sync.

Reply:
> I don't follow this: what's at the top then?

The latest instruction in the current context (current signal).

> I'm thinking we should have a single stack with a compound data structure,

That does seem simpler but it may create some more complexities IMHO: we'd need to look at two entries of the stack for the operation here (since we want the second-to-top for pre_signal_instr_). Also updating the top instr (replace_top) will become a bit more expensive since it's a bigger entry.

> we have to worry about being out of sync.

We push/pop all at the same time always. 

If you feel strongly about this, I can try changing it. But I don't see a clear win in clarity.

Reply:
> The latest instruction in the current context (current signal).

But isn't that the current record?  So this is just an ordering problem where it's already set the prev to equal the current?

Reply:
> We push/pop all at the same time always.

This seems error-prone.

> If you feel strongly about this, I can try changing it. But I don't see a clear win in clarity.

Yes I do think a single stack is a better solution.  The code is simpler and clearer; the invariants are clearer (no more multiple comments and asserts about how all these separate stacks are supposed to be coordinated); changing an aspect of the top of the stack looks simpler (set a field, instead of pop;push).

Reply:
Thinking more on this:  I think this can be simplified further. I'll add something like `last_instr_in_cur_context` which can repeatedly be updated, and then keep all context for the current signal in the same entry on one stack.

Reply:
Done.